File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/RepositoryGenerator.java
package ru.tinkoff.kora.database.annotation.processor;

import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeSpec;

import javax.annotation.Nullable;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeMirror;

public interface RepositoryGenerator {
    TypeSpec generate(TypeElement typeElement, TypeSpec.Builder type, MethodSpec.Builder constructor);

    @Nullable
    // todo TypeName
    TypeMirror repositoryInterface();
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/RepositoryGenerator.kt
package ru.tinkoff.kora.database.symbol.processor

import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.TypeSpec

interface RepositoryGenerator {
    fun generate(
        repositoryType: KSClassDeclaration,
        typeBuilder: TypeSpec.Builder,
        constructorBuilder: FunSpec.Builder
    ): TypeSpec

    fun repositoryInterface(): KSType?
}
